scapy/contrib/iec104_information_elements.py

```diff
@@ -0,0 +1,1356 @@
+
+        This module is distributed in the hope that it will be useful,
+        but WITHOUT ANY WARRANTY; without even the implied warranty of
+        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+        GNU General Public License for more details.
+
+    :description:
+
+        This module provides the information element (IE) definitions used to compose the
+        ASDUs (Application Service Data Units) used within the IEC 60870-5-101 and IEC 60870-5-104 protocol.
+
+        normative references:
+            - IEC 60870-5-4:1993 (atomic base types / data format)
+            - IEC 60870-5-101:2003 (information elements (sec. 7.2.6) and ASDU definition (sec. 7.3))
+            - IEC 60870-5-104:2006 (information element TSC (sec. 8.8, p. 44))
+
+    :TODO:
+        - some definitions should use signed types as outlined in the standard
+        - normed value element should use a float type
+
+    :NOTES:
+        - some of the information element definitions are very verbose - sometimes the whole range of a byte
+          is used to define reserved value ranges <--> ranges are added explicitly in 'full text' to allow
+          simple parsing of the definitions for generating other representations (in other words: those  lengthy
+          flag definitions are there 'on purpose')
+        -
+"""
+from scapy.contrib.iec104_fields import LEIEEEFloatField, IEC60870_5_4_NormalizedFixPoint, LESignedShortField, \
+    IEC104SignedSevenBitValue
+from scapy.fields import BitEnumField, ByteEnumField, ByteField, ThreeBytesField, \
+    BitField, LEShortField, LESignedIntField
+from scapy.modules.six.moves import range
+
+
+def _generate_attributes_and_dicts(cls):
+    """
+    create class attributes and dict entries for range-based attributes
+
+    class attributes will take the form: cls.<attribute_name_prefix>_<index>
+
+    dictionary entries will be generated as: the_dict[index] = "<dict_entry_prefix> (<index>)"
+
+    expects a GENERATED_ATTRIBUTES attribute withing the class that contains a list of the specification for the
+    attributes and dictionary entries to be generated. each list entry must have this format:
+
+        (attribute_name_prefix, dict_entry_prefix, dictionary, first_index, last_index)
+
+    with
+        <attribute_name_prefix> - the prefix of the attribute name
+        first_index - index of the first attribute to be generated
+        last_index - index of the last attribute to be generated
+    :param cls: the class the attributes should be added to
+    :return: cls extended by generated attributes
+    """
+
+    for attribute_name_prefix, dict_entry_prefix, the_dict, first_index, last_index in cls.GENERATED_ATTRIBUTES:
+
+        for index in range(first_index, last_index + 1):
+
+            the_dict[index] = '{} ({})'.format(dict_entry_prefix, index)
+
+            setattr(cls, '{}_{}'.format(attribute_name_prefix, index), index)
+
+    return cls
+
+
+class IEC104_IE_CommonQualityFlags:
```

Comment:
The last problem I have (thanks for all the fixes BTW :-) ) is that you have some non-useable classes such as this one that use the same naming convention as the packets.

It’s unclear if you can call them or not. If they are restricted to an internal usage, add a _ before. If they are expected to be used by a random user, maybe find a different naming convention ?

Reply:
The classes in information_elements are the building-blocks for the information_objects - they deliver the composed information objects fields (therefore the fields_desc) and related element attribute value definitions. 

As the user will access the classes directly to get the attr value definitions the classes should stay public.

Example:
`IEC104_I_Message_SeqIOA(io=IEC104_IO_M_DP_NA_1(dpi_value=IEC104_IE_DIQ.DPI_FLAG_STATE_UNDEFINED))`

IEC104_IE_DIQ is one of those information elements in question.

I agree that using fields_desc is misleading - renamed it to informantion_element_fields in bf27f99.

Is this sufficient?

Reply:
Great thanks !